File: install_packages.jl
using Pkg

PACKAGES = [
    "CSV", "DataFrames", "GeoDataFrames", "ArchGDAL", "Statistics", "LibGEOS",
    "CategoricalArrays", "Printf", "Random", "Mice", "GLM",
    "CovarianceMatrices", "Serialization", "LinearAlgebra", "Distributions",
    "CairoMakie", "Latexify", "GeoInterfaceMakie", "StatsBase", "ZipFile"
]

"""
    find_missing(pkg_list)

Return the subset of pkg_list that are not currently installed.
Uses Pkg.project() to check against the active environment's dependencies,
with a fallback check against the full depot for packages installed globally.

# Arguments
- `pkg_list::Vector{String}`: package names to check.

# Returns
- `Vector{String}`: names of packages not found in the environment.
"""
function find_missing(pkg_list::Vector{String})::Vector{String}
    installed_names = Set(v.name for v in values(Pkg.dependencies()))
    return [pkg for pkg in pkg_list if pkg ∉ installed_names]
end


"""
    report_status(pkg_list)

Print the installation status of every package in pkg_list.

# Arguments
- `pkg_list::Vector{String}`: package names to check.
"""
function report_status(pkg_list::Vector{String})
    println("Checking installed Julia packages...")
    installed = Set(v.name for v in values(Pkg.dependencies()))
    for pkg in pkg_list
        if pkg ∈ installed
            println("  $pkg is already installed")
        else
            println("  $pkg - MISSING")
        end
    end
end


"""
    install_and_verify(pkg_list)

Report current status, install any missing packages, then verify.

# Arguments
- `pkg_list::Vector{String}`: package names to check and install.
"""
function install_and_verify(pkg_list::Vector{String})
    report_status(pkg_list)

    missing_pkgs = find_missing(pkg_list)

    if isempty(missing_pkgs)
        println("\nNo missing packages :D")
        return
    end

    println("\nFound $(length(missing_pkgs)) missing package(s). Installing...")

    for pkg in missing_pkgs
        try
            Pkg.add(pkg)
            println("  Successfully installed: $pkg")
        catch e
            println("  Failed to install: $pkg ($(sprint(showerror, e)))")
        end
    end

    println("\nVerifying installation...")
    still_missing = find_missing(pkg_list)

    if !isempty(still_missing)
        println("\nThe following packages failed to install or load:")
        for pkg in still_missing
            println("  - $pkg")
        end
    else
        println("\nAll packages are now properly installed and ready to use!")
    end
end

install_and_verify(PACKAGES)

File: install_packages.py
import subprocess
import sys


PACKAGES = [
    ("pathlib",         "pathlib"),
    ("pandas",          "pandas"),
    ("geopandas",       "geopandas"),
    ("shapely",         "shapely"),
    ("time",            "time"),
    ("osmium",          "osmium"),
    ("multiprocessing", "multiprocessing"),
    ("miceforest",      "miceforest"),
    ("numpy",           "numpy"),
    ("scipy",           "scipy"),
    ("statsmodels",     "statsmodels"),
    ("re",              "re"),
    ("pygris",          "pygris"),
    ("warnings",        "warnings"),
    ("matplotlib",      "matplotlib"),
    ("seaborn",         "seaborn"),
]

def find_missing(pkg_list):
    """
    Check which packages from a list cannot be imported.

    Parameters
    ----------
    pkg_list : list of tuple
        Each tuple is (install_name, import_name).

    Returns
    -------
    list of tuple
        Subset of pkg_list whose import_name cannot be imported.
    """
    missing = []
    for install_name, import_name in pkg_list:
        try:
            __import__(import_name)
        except ImportError:
            missing.append((install_name, import_name))
    return missing


def report_status(pkg_list):
    """
    Print the installation status of every package in pkg_list.

    Parameters
    ----------
    pkg_list : list of tuple
        Each tuple is (install_name, import_name).
    """
    print("Checking installed Python packages...")
    for install_name, import_name in pkg_list:
        try:
            __import__(import_name)
            print(f"  {install_name} is already installed")
        except ImportError:
            print(f"  {install_name} - MISSING")


def install_and_verify(pkg_list):
    """
    Report current status, install any missing packages, then verify.

    Parameters
    ----------
    pkg_list : list of tuple
        Each tuple is (install_name, import_name).
    """
    report_status(pkg_list)

    missing = find_missing(pkg_list)

    if not missing:
        print("\nNo missing packages owo")
        return

    print(f"\nFound {len(missing)} missing package(s). Installing...")

    for install_name, import_name in missing:
        try:
            subprocess.check_call(
                [sys.executable, "-m", "pip", "install", install_name],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            print(f"  Successfully installed: {install_name}")
        except subprocess.CalledProcessError as e:
            print(f"  Failed to install: {install_name} ({e})")

    # Final verification pass after installation attempts
    print("\nVerifying installation...")
    still_missing = find_missing(pkg_list)

    if still_missing:
        names = [n for n, _ in still_missing]
        print(f"\nThe following packages failed to install or load: {names}")
    else:
        print("\nAll packages are now properly installed and ready to use!")

if __name__ == "__main__":
    install_and_verify(PACKAGES)